File: src/dk/eightyplus/Painter/fragment/NoteListAdapter.java
package dk.eightyplus.Painter.fragment;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import dk.eightyplus.Painter.R;
import dk.eightyplus.Painter.utilities.Compatibility;

import java.lang.ref.SoftReference;
import java.util.HashMap;
import java.util.Map;

/**
 * User: fries
 * Date: 29/03/14
 * Time: 19.34
 */
public class NoteListAdapter extends ArrayAdapter<String> implements ThumbLoader.Notify {

  @SuppressWarnings("unused")
  private static final String TAG = NoteListAdapter.class.toString();

  private static final int initialThumbLoadSize = 3;
  private Map<String, SoftReference<Drawable>> cachedImages = new HashMap<String, SoftReference<Drawable>>();
  private Map<String, Boolean> loading = new HashMap<String, Boolean>();

  private ButtonOnClickListener buttonOnClickListener;

  public NoteListAdapter(Context context, ListView listView, int resource, String[] objects) {
    super(context, resource, objects);
    for (int i = 0; i < initialThumbLoadSize && i < objects.length; i++) {
      startImageLoad(objects[i]);
    }

    buttonOnClickListener = new ButtonOnClickListener(listView);
  }

  @Override
  public boolean hasStableIds() {
    return true;
  }

  @Override
  public View getView(int position, View convertView, ViewGroup parent) {
    if (convertView == null) {
      LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
      convertView = inflater.inflate(R.layout.note_list_item, parent, false);
    }
    View rowView = convertView;

    TextView textView = (TextView) rowView.findViewById(R.id.note_title);
    final ImageView imageView = (ImageView) rowView.findViewById(R.id.note_thumb);
    String item = getItem(position);
    textView.setText(item);

    final View buttonDelete = rowView.findViewById(R.id.button_delete);
    buttonDelete.setTag(R.id.button_delete);
    buttonDelete.setOnClickListener(buttonOnClickListener);

    if (!setImage(imageView, item)) {
      imageView.setImageResource(android.R.drawable.stat_notify_error);
      startImageLoad(item);
    }

    return rowView;
  }

  public void onPause() { }

  public void onResume() { }

  public void onDestroy() { }

  private boolean setImage(ImageView imageView, String path) {
    if (cachedImages.containsKey(path)) {
      Drawable drawable = cachedImages.get(path).get();
      if (drawable != null) {
        imageView.setImageDrawable(drawable);
        return true;
      }
    }
    return false;
  }

  public Drawable getImage(int position) {
    String path = getItem(position);
    if (cachedImages.containsKey(path)) {
      return cachedImages.get(path).get();
    }
    return null;
  }

  private void startImageLoad(String s) {
    if (doStartLoad(s)) {
      setLoadingStarted(s);
      Compatibility.get().startTask(new ThumbLoader(getContext(), this), s);
    }
  }

  private void setLoadingStarted(String key) {
    loading.put(key, true);
  }

  private void setLoadingDone(String key) {
    loading.remove(key);
  }

  private boolean doStartLoad(String s) {
    return !loading.containsKey(s) || loading.get(s) == false;
  }

  @Override
  public void done(String path, Drawable drawable) {
    cachedImages.put(path, new SoftReference<Drawable>(drawable));
    setLoadingDone(path);
    notifyDataSetChanged();
  }

  /**
   * Small class to put on buttons inside list view item, passed click with id of button
   */
  private static class ButtonOnClickListener implements View.OnClickListener {
    private final SoftReference<ListView> listViewSoftReference;

    public ButtonOnClickListener(ListView listView) {
      this.listViewSoftReference = new SoftReference<ListView>(listView);
    }

    @Override
    public void onClick(View v) {
      ListView listView = listViewSoftReference.get();
      if (listView != null) {
        int position = listView.getPositionForView((View) v.getParent());
        listView.performItemClick(v, position, (Integer) v.getTag());
      }
    }
  };
}

File: app/src/main/java/dk/eightyplus/annoto/fragment/NoteListAdapter.kt
package dk.eightyplus.annoto.fragment

import android.content.Context
import android.graphics.drawable.Drawable
import android.os.AsyncTask
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import dk.eightyplus.annoto.R
import dk.eightyplus.annoto.utilities.Compatibility

import java.lang.ref.SoftReference
import java.util.HashMap

/**
 * User: fries
 * Date: 29/03/14
 * Time: 19.34
 */
class NoteListAdapter(context: Context, listView: ListView, resource: Int, objects: Array<String>) : ArrayAdapter<String>(context, resource, objects), ThumbLoader.Notify {
    private val cachedImages = HashMap<String, SoftReference<Drawable>>()
    private val loading = HashMap<String, Boolean>()

    private val buttonOnClickListener: ButtonOnClickListener
    private val initialThumbLoadSize = 3

    init {
        var i = 0
        while (i < initialThumbLoadSize && i < objects.size) {
            startImageLoad(objects[i])
            i++
        }

        buttonOnClickListener = ButtonOnClickListener(listView)
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val rowView = if (convertView == null) {
            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            inflater.inflate(R.layout.note_list_item, parent, false)
        } else {
            convertView
        }

        val textView = rowView.findViewById<View>(R.id.note_title) as TextView
        val imageView = rowView.findViewById<View>(R.id.note_thumb) as ImageView
        val item = getItem(position)
        textView.text = item

        val buttonDelete = rowView.findViewById<View>(R.id.button_delete)
        buttonDelete.tag = R.id.button_delete
        buttonDelete.setOnClickListener(buttonOnClickListener)

        if (!setImage(imageView, item)) {
            imageView.setImageResource(android.R.drawable.stat_notify_error)
            startImageLoad(item)
        }

        return rowView
    }

    fun onPause() {}

    fun onResume() {}

    fun onDestroy() {}

    private fun setImage(imageView: ImageView, path: String?): Boolean {
        if (cachedImages.containsKey(path)) {
            cachedImages[path]?.get().let {
                imageView.setImageDrawable(it)
                return true
            }
        }
        return false
    }

    fun getImage(position: Int): Drawable? {
        val path = getItem(position)
        return if (cachedImages.containsKey(path)) {
            cachedImages[path]?.get()
        } else null
    }

    private fun startImageLoad(s: String) {
        if (doStartLoad(s)) {
            setLoadingStarted(s)
            Compatibility.get().startTask(ThumbLoader(context, this), s)
        }
    }

    private fun setLoadingStarted(key: String) {
        loading[key] = true
    }

    private fun setLoadingDone(key: String) {
        loading.remove(key)
    }

    private fun doStartLoad(s: String?): Boolean {
        return !loading.containsKey(s) || loading[s] == false
    }

    override fun done(path: String, drawable: Drawable) {
        cachedImages[path] = SoftReference(drawable)
        setLoadingDone(path)
        notifyDataSetChanged()
    }

    /**
     * Small class to put on buttons inside list view item, passed click with id of button
     */
    private class ButtonOnClickListener(listView: ListView) : View.OnClickListener {
        private val listViewSoftReference: SoftReference<ListView> = SoftReference(listView)

        override fun onClick(v: View) {
            listViewSoftReference.get()?.let {
                val position = it.getPositionForView(v.parent as View)
            it.performItemClick(v, position, (v.tag as Int).toLong())
            }
        }
    }
}
